webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/write/PolicyWriterService.scala

```diff
@@ -512,58 +568,76 @@ class PolicyWriterServiceImpl(
       template
     }
 
-    val now = System.currentTimeMillis()
-
     /*
      * NOTE : this is inefficient and store in a lot of multiple time the same content
      * if only the outpath change for two differents agent type.
      */
-   val res = (parrallelSequence(templatesToRead) { case (templateId, agentType, templateOutPath) =>
-      for {
-        copyInfo <- techniqueRepository.getTemplateContent(templateId) { optInputStream =>
-          optInputStream match {
-            case None =>
-              Failure(s"Error when trying to open template '${templateId.toString}'. Check that the file exists with a ${TechniqueTemplate.templateExtension} extension and is correctly commited in Git, or that the metadata for the technique are corrects.")
-            case Some(inputStream) =>
-              logger.trace(s"Loading template: ${templateId}")
-              //string template does not allows "." in path name, so we are force to use a templateGroup by polity template (versions have . in them)
-              val content = IOUtils.toString(inputStream, StandardCharsets.UTF_8)
-              Full(TechniqueTemplateCopyInfo(templateId, templateOutPath, content))
-          }
-        }
-      } yield {
-        ((copyInfo.id, agentType), copyInfo)
-      }
-    }).map( _.toMap)
-
-    logger.debug(s"${templatesToRead.size} promises templates read in ${System.currentTimeMillis-now} ms")
-    res
+    for {
+      t0  <- currentTimeMillis
+      res <- (parrallelSequence(templatesToRead) { case (templateId, agentType, templateOutPath) =>
+               for {
+                 copyInfo <- techniqueRepository.getTemplateContent(templateId) { optInputStream =>
+                               optInputStream match {
+                                 case None =>
+                                   Unexpected(s"Error when trying to open template '${templateId.toString}'. Check that the file exists with a ${TechniqueTemplate.templateExtension} extension and is correctly commited in Git, or that the metadata for the technique are corrects.").fail
+                                 case Some(inputStream) =>
+                                   for {
+                                     _       <- PolicyLoggerPure.trace(s"Loading template: ${templateId}")
+                                               //string template does not allows "." in path name, so we are force to use a templateGroup by polity template (versions have . in them)
+                                     content <- IOResult.effect(s"Error when copying technique template '${templateId.toString}'")(inputStream.asString(false))
```

Comment:
is switching from IOUtils to better.files as a noticable impact ? if so, we could do that in 5.0 also

Reply:
so it seems that performance are no better